AwsCryptographicMaterialProviders/runtimes/java/src/testExamples/java/software/amazon/cryptography/example/hierarchy/concurrent/StorageVersionReadConcurrencyTests.java

```diff
@@ -0,0 +1,302 @@
+    res
+      .items()
+      .forEach(response -> {
+        DdbHelper.deleteKeyStoreDdbItem(
+          branchKeyId,
+          response.get("type").s(),
+          Fixtures.TEST_KEYSTORE_NAME,
+          DynamoDbClient.create(),
+          true
+        );
+      });
+  }
+
+  private Storage storageForThread(final String threadIdToIndex) {
+    return threadIndexToStorage.computeIfAbsent(
+      threadIdToIndex,
+      k -> createStorageLayer()
+    );
+  }
+
+  private KeyStore keyStoreForThread(String threadIdToIndex) {
+    return threadIndexToKeyStore.computeIfAbsent(
+      threadIdToIndex,
+      k -> createKeyStore()
+    );
+  }
+
+  private void raceToVersionWithStorage(KeyStoreAdmin admin) {
+    VersionKeyInput input = VersionKeyInput
+      .builder()
+      .Identifier(branchKeyId)
+      .KmsArn(
+        KmsSymmetricKeyArn
+          .builder()
+          .KmsKeyArn(Fixtures.KEYSTORE_KMS_ARN)
+          .build()
+      )
+      .build();
+    admin.VersionKey(input);
+  }
+
+  private GetActiveBranchKeyOutput raceToReadActiveWithKeyStore(
+    KeyStore keyStore
+  ) {
+    GetActiveBranchKeyInput input = GetActiveBranchKeyInput
+      .builder()
+      .branchKeyIdentifier(branchKeyId)
+      .build();
+    return keyStore.GetActiveBranchKey(input);
+  }
+
+  @Test(threadPoolSize = 10, invocationCount = 100, timeOut = 10000)
+  public void testConcurrentVersionWithStorage() {
+    String threadId = String.valueOf(Thread.currentThread().getId());
+    String threadIdToIndex = storageIndexToThreadId.computeIfAbsent(
+      threadId,
+      str -> unpickedIndicesForStorage.pop()
+    );
+
+    String timestamp = dateFormat.format(new Date());
+    try {
+      Storage threadStorage = storageForThread(threadIdToIndex);
+      KeyStoreAdminConfig keyStoreAdminConfig = KeyStoreAdminConfig
+        .builder()
+        .storage(threadStorage)
+        .logicalKeyStoreName(Fixtures.TEST_KEYSTORE_NAME)
+        .build();
+      KeyStoreAdmin admin = KeyStoreAdmin
+        .builder()
+        .KeyStoreAdminConfig(keyStoreAdminConfig)
+        .build();
+      raceToVersionWithStorage(admin);
+      System.out.println(
+        "Successfully versioned branch key! Thread ID: " +
+        Thread.currentThread().getId() +
+        " ThreadIndex: " +
+        threadIdToIndex +
+        " Timestamp: " +
+        timestamp +
+        " BranchKeyId: " +
+        branchKeyId
+      );
+    } catch (TransactionCanceledException exception) {
+      System.out.println("Failed to write branch key: " + branchKeyId);
+      // Exceptions that get thrown when you write keys using the Storage interface
+      exception
+        .cancellationReasons()
+        .forEach(cancellationReason -> {
+          Assert.assertTrue(
+            (cancellationReason.code().equals("TransactionConflict") ||
+              cancellationReason.code().equals("None") ||
+              cancellationReason.code().equals("ConditionalCheckFailed"))
```

Comment:
This assert only fires if an Exception is thrown.

There is no assert on the non-catch path.

Also, there is no evidence if, in the 10 * 100 = 1000 invocations,
one of these asserts was met.

Reply:
> There is no assert on the non-catch path.

Correct, we only expect that in a system that is trying to race to version, these would be the only errors that get thrown. Any other error should fail the test.